src/main/host/syscall_types.rs

```diff
@@ -89,51 +89,78 @@ impl PartialEq for SysCallReg {
 
 impl Eq for SysCallReg {}
 
-impl From<u64> for SysCallReg {
-    fn from(v: u64) -> Self {
-        Self { as_u64: v }
-    }
-}
-
 impl From<SysCallReg> for u64 {
     fn from(v: SysCallReg) -> u64 {
         unsafe { v.as_u64 }
     }
 }
 
-impl From<u32> for SysCallReg {
-    fn from(v: u32) -> Self {
-        Self { as_u64: v as u64 }
+impl From<SysCallReg> for i64 {
+    fn from(v: SysCallReg) -> i64 {
+        unsafe { v.as_i64 }
     }
 }
 
-impl From<SysCallReg> for u32 {
-    fn from(v: SysCallReg) -> u32 {
-        (unsafe { v.as_u64 }) as u32
-    }
+macro_rules! try_from_syscallreg {
+    ($to_type:ty, $using_field:ident) => {
+        impl TryFrom<SysCallReg> for $to_type {
+            //type Error = <Self as TryFrom<u64>>::Error;
+            type Error = nix::errno::Errno;
+
+            fn try_from(v: SysCallReg) -> Result<Self, Self::Error> {
+                let val = unsafe { v.$using_field };
+
```

Comment:
I think this could just be something like `val.try_into().map_err(|_| Self::Error::EINVAL)`?

In particular I'm not sure what you have now will work properly with signed integers; I think e.g. a -1 passed through a 32-bit int at the library level might get sign extended into a 64-bit -1 with all 1-bits for the syscall.

Maybe worth some unit tests either way?

Reply:
glibc doesn't sign extend them before making the syscall, so this doesn't work for signed types. But I guess other libc implementations could sign extend them.

Reply:
Hmm, yeah I can see how it ends up not getting sign extended, though it's not clear to me whether we can depend on it never being sign extended.

Maybe allow the high bits to be all 1's if the converted value is signed and negative?

Reply:
In any case, worth documenting why we don't do the simpler thing here and what we do and don't know about the arguments being sign-extended :)